### PR TITLE
Libretro

### PR DESCRIPTION
**Pull request by Kim-Dewelski**
a WIP pull request for libretro support. Not ready to be merged yet.

### CODE REVIEW

File: src/libretro.h [hunk 1]
```diff
@@ -0,0 +1,4405 @@
+/* Copyright (C) 2010-2020 The RetroArch team
```
Comment (skylersaleh) [inline src/libretro.h:1]: Did you make any changes to this file for the SkyEmu port? Or is it just the straight header from the retro arch project?

File: src/libretro.c [hunk 2]
```diff
@@ -0,0 +1,843 @@
...
+#include <string.h>
+#include <math.h>
+
+#include "shared.h"
+#include "sb_types.h"
+#include "gb.h"
+#include "gba.h"
+#include "nds.h"
+
+#define SKYEMU_LIBRETRO_VERSION "0.1.0"
```
Comment (skylersaleh) [inline src/libretro.c:14]: Can this be based on the GIT_COMMIT_HASH define that we set in our CMake file?

That way we can correlate it back with the regular builds and have it automatically update.

File: src/libretro.c [hunk 3]
```diff
@@ -0,0 +1,843 @@
...
+
+// necessary to resolve a linker error when loaded by retroarch
+bool se_load_bios_file(const char* name, const char* base_path, const char* file_name, uint8_t* data, size_t data_size) {
```
Comment (skylersaleh) [inline src/libretro.c:17]: Its probably better to make this a callback function in the cores that you can just set to NULL in the retro arch load rom commands. That way the cores are completely decoupled from the frontend.

File: src/libretro.c [hunk 4]
```diff
@@ -0,0 +1,843 @@
...
+static retro_input_poll_t input_poll_cb = NULL;
+static retro_input_state_t input_state_cb = NULL;
+static retro_audio_sample_t audio_sample_cb = NULL;
+static retro_environment_t env_cb = NULL;
+
+sb_emu_state_t emu_state;
+
+/* ------------------ RETROARCH GB ----------------- */
+
+static gb_scratch_t gb_scratch;
```
Comment (skylersaleh) [inline src/libretro.c:36]: You should union the cores/scratch space for all the systems like the main frontend to save RAM.

File: src/libretro.c [hunk 5]
```diff
@@ -0,0 +1,843 @@
...
+  return sb_load_rom(&emu_state, &gb, &gb_scratch);
+}
+
+void retro_gb_get_system_av_info(struct retro_system_av_info* info) {  
+  info->geometry.aspect_ratio = 0.0;
+  info->geometry.max_height = SB_LCD_H;
+  info->geometry.max_width = SB_LCD_W;
+  info->geometry.base_height = info->geometry.max_height;
+  info->geometry.base_width = info->geometry.max_width;
+  info->timing.fps = 60;
```
Comment (skylersaleh) [inline src/libretro.c:117]: 59.727 FPS

File: src/libretro.c [hunk 6]
```diff
@@ -0,0 +1,843 @@
...
+  return gba_load_rom(&emu_state, &gba, &gba_scratch);
+}
+
+void retro_gba_get_system_av_info(struct retro_system_av_info* info) {
+  info->geometry.aspect_ratio = 0.0;
+  info->geometry.max_width = GBA_LCD_W;
+  info->geometry.max_height = GBA_LCD_H;
+  info->geometry.base_width = info->geometry.max_width;
+  info->geometry.base_height = info->geometry.max_height; 
+  info->timing.fps = 60;
```
Comment (skylersaleh) [inline src/libretro.c:275]: 59.727

File: src/libretro.c [hunk 7]
```diff
@@ -0,0 +1,843 @@
...
+  return nds_load_rom(&emu_state, &nds, &nds_scratch);
+}
+
+void retro_nds_get_system_av_info(struct retro_system_av_info* info) {
+  info->geometry.aspect_ratio = 0.0;
+  info->geometry.max_width = NDS_LCD_W;
+  info->geometry.max_height = NDS_LCD_H * 2; // (bottom and top screen)
+  info->geometry.base_width = info->geometry.max_width;
+  info->geometry.base_height = info->geometry.max_height; 
+  info->timing.fps = 60;
```
Comment (skylersaleh) [inline src/libretro.c:514]: 59.8261

**Review by skylersaleh — CHANGES_REQUESTED** [review]

File: src/libretro.h (hunk 1 above)
Comment (Kim-Dewelski) [reply src/libretro.h:1]: it is a straight copy

File: src/libretro.c (hunk 2 above)
Comment (Kim-Dewelski) [reply src/libretro.c:14]: sure, I'll look into that

File: src/libretro.c (hunk 3 above)
Comment (Kim-Dewelski) [reply src/libretro.c:17]: I actually quite like how this ended up, I made some changes to this function (that I have yet pushed), so this function is actually used at this point and I don't know if we should remove it still, in favour of a callback?

Comment (skylersaleh) [reply src/libretro.c:17]: Lets push your set of changes and go from there!

**Comment by skylersaleh** [comment]
One observation is that if you drag from the bottom touch screen to a location overlapping the top touch screen of the NDS, it yields erratic touch points. I assume this is a result of not sanitizing the touch coordinates to only the region of the bottom touch screen.

I also can't seem to get cheat codes to work correctly. Have you tested them out?

**Comment by Kim-Dewelski** [comment]
What cheat codes are not working properly for you?

**Comment by skylersaleh** [comment]
Can you resolve the merge conflict please?

**Comment by skylersaleh** [comment]
image

**Pull request merged**